### PR TITLE
Add server-side match "rulesets" (including Team VS)

### PR DESCRIPTION
**Pull request by peppy**
This is in a pretty usable state, except for two things:

- I highly recommend we rename `MatchRuleset` to *anything else*.
- We need to update osu-web / database flow to handle the inital room creation case. Right now you can only change to Team VS after already starting a room in the default (head-to-head) mode.

Also,

- [x] Depends on https://github.com/ppy/osu/pull/14104

I welcome review of the general structure of this while we figure out the last pieces of the puzzle.

### CODE REVIEW

**Review by bdach — CHANGES_REQUESTED** [review]

File: osu.Server.Spectator.Tests/JsonSerializationTests.cs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+            };
+
+            var serialized = JsonConvert.SerializeObject(state, typeof(MatchRulesetUserState), settings);
+
+            var deserializedState = JsonConvert.DeserializeObject(serialized, settings);
+            var deserializedRoomState = deserializedState as TeamVsMatchUserState;
+
+            Assert.NotNull(deserializedRoomState);
+
+            Assert.Equal(state.TeamID, deserializedRoomState!.TeamID);
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Server.Spectator.Tests/JsonSerializationTests.cs:34]: i don't know, you tell me. the assert above should in my opinion be enough. i just threw fire at this to make it compile. my opinion has not changed, i just gave up trying to figure out how to appease the compiler.

File: osu.Server.Spectator/Hubs/HeadToHeadRuleset.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+            // we don't need a state, but keep things simple by completely nulling the state.
+            // this allows the client to see a user state change and handle match ruleset specifics based on that alone.
+            user.MatchRulesetState = null;
+            Room.UpdateMatchRulesetUserState(Room, user);
+        }
+
+        public override void HandleUserRequest(MultiplayerRoomUser user, MatchRulesetUserRequest request)
+        {
+        }
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Server.Spectator/Hubs/HeadToHeadRuleset.cs:27]: Was abstract at some point, can probably be removed.

File: osu.Server.Spectator.Tests/JsonSerializationTests.cs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        {
+            MultiplayerRoom room = new MultiplayerRoom(1234)
+            {
+                Users =
+                {
+                    new MultiplayerRoomUser(888),
+                }
+            };
+
+            var serialized = JsonConvert.SerializeObject(room, typeof(MatchRulesetRoomState), settings);
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Server.Spectator.Tests/JsonSerializationTests.cs:73]: yes. without it the serialisation will as you would expect, fail. unless the serialisation name handling mode is set to `All`.

File: osu.Server.Spectator/Hubs/TeamVsRuleset.cs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+            room.UpdateMatchRulesetRoomState(room);
+        }
+
+        public override void HandleUserJoined(MultiplayerRoomUser user)
+        {
+            base.HandleUserJoined(user);
+
+            // if the state already has a team specification we can reuse it (and keep the user's chosen team).
+            if (user.MatchRulesetState is TeamVsMatchUserState)
+                return;
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Server.Spectator/Hubs/TeamVsRuleset.cs:28]: Probably yeah. The idea was to show that this could potentially retain team choice between ruleset selection, but I'll just remove for now as it's not required for any existing functionality.

File: osu.Server.Spectator.Tests/JsonSerializationTests.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (bdach) [reply osu.Server.Spectator.Tests/JsonSerializationTests.cs:34]: The `.AsNonNull()` thing we usually do *should* work here I think? It's defined in framework which is transitively referenced (although whether it should remain to be in the future may be a point of contention?)

File: osu.Server.Spectator/Hubs/TeamVsRuleset.cs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+        /// <summary>
+        /// For a user joining the room, this will provide the most appropriate team for the new user to keep the room balanced.
+        /// </summary>
+        private int getBestAvailableTeam()
+        {
+            // initially check for any teams which don't yet have players, but are lower than TeamCount.
+            foreach (var team in state.Teams)
+            {
+                if (Room.Users.Count(u => (u.MatchRulesetState as TeamVsMatchUserState)?.TeamID == team.ID) == 0)
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Server.Spectator/Hubs/TeamVsRuleset.cs:58]: Has to become `.All` due to inspections, which makes it not really faster but maybe more legible.

File: osu.Server.Spectator/Hubs/MultiplayerHub.cs [hunk 6]
```diff
@@ -98,11 +98,14 @@ public async Task<MultiplayerRoom> JoinRoomWithPassword(long roomId, string pass
                         }
 
                         userUsage.Item = new MultiplayerClientState(Context.ConnectionId, CurrentContextUserId, roomId);
+                        // because match rulesets may send subsequent information via Users collection hooks,
+                        // inform clients before adding user to the room.
+                        await Clients.Group(GetGroupId(roomId)).UserJoined(roomUser);
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Server.Spectator/Hubs/MultiplayerHub.cs:103]: Actually the database already catches and eats all errors, so this isn't a huge concern. Also out of scope of this PR as it hasn't been moved here.

File: osu.Server.Spectator/Hubs/TeamVsRuleset.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (bdach) [reply osu.Server.Spectator/Hubs/TeamVsRuleset.cs:58]: It's still theoretically faster even if it's `All()`, as that will bail on the first non-matching item (as opposed to `Count()` which always enumerates to end).

[the remark this answers is not in the document]
Comment (peppy) [reply osu.Server.Spectator/Hubs/TeamVsRuleset.cs]: Fair.

File: osu.Server.Spectator.Tests/JsonSerializationTests.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Server.Spectator.Tests/JsonSerializationTests.cs]: I've updated to use that.

**Comment by peppy** [comment]
@smoogipoo I've applied the structural changes we discussed. Let me know if I've missed anything.

File: osu.Server.Spectator/Hubs/ServerMultiplayerRoom.cs [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    public class ServerMultiplayerRoom : MultiplayerRoom
+    {
+        private readonly IMultiplayerServerMatchCallbacks hubCallbacks;
+
+        private MatchTypeImplementation matchTypeImplementation;
+
+        public MatchTypeImplementation MatchTypeImplementation
+        {
+            get => matchTypeImplementation;
+            set
```
Comment (smoogipoo) [inline osu.Server.Spectator/Hubs/ServerMultiplayerRoom.cs:18]: probably should be `private set;`

File: SampleMultiplayerClient/MultiplayerClient.cs [hunk 8]
```diff
@@ -37,6 +37,8 @@ public MultiplayerClient(HubConnection connection, int userId)
             connection.On(nameof(IMultiplayerClient.MatchStarted), ((IMultiplayerClient)this).MatchStarted);
             connection.On(nameof(IMultiplayerClient.ResultsReady), ((IMultiplayerClient)this).ResultsReady);
             connection.On<int, IEnumerable<APIMod>>(nameof(IMultiplayerClient.UserModsChanged), ((IMultiplayerClient)this).UserModsChanged);
+            connection.On<MatchRoomState>(nameof(IMultiplayerClient.MatchRoomStateChanged), ((IMultiplayerClient)this).MatchRoomStateChanged);
+            connection.On<int, MatchUserState>(nameof(IMultiplayerClient.MatchUserStateChanged), ((IMultiplayerClient)this).MatchUserStateChanged);
```
Comment (smoogipoo) [inline SampleMultiplayerClient/MultiplayerClient.cs:41]: Handle `MatchEvent` here too

File: SampleMultiplayerClient/MultiplayerClient.cs [hunk 9]
```diff
@@ -129,6 +136,21 @@ Task IMultiplayerClient.UserStateChanged(int userId, MultiplayerUserState state)
...
+
+        public Task MatchRoomStateChanged(MatchRoomState state)
+        {
+            throw new NotImplementedException();
+        }
+
+        public Task MatchEvent(MatchServerEvent e)
+        {
+            throw new NotImplementedException();
+        }
```
Comment (smoogipoo) [inline SampleMultiplayerClient/MultiplayerClient.cs:152]: Maybe implement? You've at least implemented changing the type in `Program.cs`.

File: SampleMultiplayerClient/Program.cs [hunk 10]
```diff
@@ -75,8 +77,16 @@ public static async Task Main()
                             await targetClient.TransferHost(int.Parse(args[0]));
                             break;
 
+                        case "changeruleset":
```
Comment (smoogipoo) [inline SampleMultiplayerClient/Program.cs:80]: `changematchtype`, probably?

**Review by smoogipoo — APPROVED** [review]
Seems fine otherwise for an initial pass

File: SampleMultiplayerClient/MultiplayerClient.cs (hunk 9 above)
Comment (peppy) [reply SampleMultiplayerClient/MultiplayerClient.cs:152]: I kind of gave up on this because I found it pretty non-useful for testing due to database requirements (plan to redo this at a later stage). I'll add the bare minimum though untested.

File: osu.Server.Spectator/Hubs/ServerMultiplayerRoom.cs (hunk 7 above)
Comment (peppy) [reply osu.Server.Spectator/Hubs/ServerMultiplayerRoom.cs:18]: I initially tried to make it private, but it is used in tests (to provide a mock implementation), else the setter method would not even be required.

File: osu.Server.Spectator/Database/DatabaseAccess.cs [hunk 11]
```diff
@@ -50,7 +50,7 @@ public async Task<bool> IsUserRestrictedAsync(int userId)
 
         public Task<multiplayer_room> GetRoomAsync(long roomId)
         {
-            return connection.QueryFirstOrDefaultAsync<multiplayer_room>("SELECT * FROM multiplayer_rooms WHERE category = 'realtime' AND id = @RoomID", new
+            return connection.QueryFirstOrDefaultAsync<multiplayer_room>("SELECT * FROM multiplayer_rooms WHERE type != 'playlists' AND id = @RoomID", new
```
Comment (peppy) [inline osu.Server.Spectator/Database/DatabaseAccess.cs:53]: This was intended to be in another PR and was actually required for https://github.com/ppy/osu-web/pull/7965#discussion_r682385215. I've deployed to production already, but will just leave it to be merged as part of this PR since it seems like it's close to ready.

**Review by smoogipoo — APPROVED** [review]
